File: app/backend/java/payment-service/src/main/java/com/srbms/paymentservice/service/PaymentService.java
package com.srbms.paymentservice.service;

import com.srbms.paymentservice.dto.*;
import com.srbms.paymentservice.entity.*;
import com.srbms.paymentservice.exception.DuplicateTransactionException;
import com.srbms.paymentservice.exception.PaymentException;
import com.srbms.paymentservice.exception.PaymentNotFoundException;
import com.srbms.paymentservice.repository.PaymentAuditLogRepository;
import com.srbms.paymentservice.repository.PaymentRepository;
import com.srbms.paymentservice.strategy.PaymentProcessorFactory;
import com.srbms.paymentservice.strategy.PaymentProcessorStrategy;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class PaymentService {

    private final PaymentRepository paymentRepository;
    private final PaymentAuditLogRepository auditLogRepository;
    private final PaymentProcessorFactory processorFactory;

    @Value("${fare.default.base:50.00}")
    private BigDecimal defaultBaseFare = new BigDecimal("50.00");

    @Value("${fare.default.per-km:15.00}")
    private BigDecimal defaultPerKmRate = new BigDecimal("15.00");

    @Autowired
    public PaymentService(PaymentRepository paymentRepository,
                          PaymentAuditLogRepository auditLogRepository,
                          PaymentProcessorFactory processorFactory) {
        this.paymentRepository = paymentRepository;
        this.auditLogRepository = auditLogRepository;
        this.processorFactory = processorFactory;
    }

    @Transactional
    public PaymentResponseDTO processPayment(PaymentRequestDTO requestDTO) {
        // 1. Idempotency Check for Ride
        if (paymentRepository.existsByRideIdAndPaymentStatus(requestDTO.getRideId(), PaymentStatus.SUCCESS)) {
            throw new DuplicateTransactionException("Payment has already been completed for Ride #" + requestDTO.getRideId());
        }

        // 2. Prepare Transaction ID
        String txnId = requestDTO.getTransactionId();
        if (txnId == null || txnId.trim().isEmpty()) {
            txnId = "TXN_" + System.currentTimeMillis() + "_" + UUID.randomUUID().toString().substring(0, 6).toUpperCase();
        } else if (paymentRepository.existsByTransactionId(txnId)) {
            throw new DuplicateTransactionException("Transaction ID already exists: " + txnId);
        }

        // 3. Construct Payment entity
        BigDecimal discount = requestDTO.getDiscountAmount() != null ? requestDTO.getDiscountAmount() : BigDecimal.ZERO;
        BigDecimal netAmount = requestDTO.getTotalFare().subtract(discount);
        if (netAmount.compareTo(BigDecimal.ZERO) < 0) {
            netAmount = BigDecimal.ZERO;
        }

        Payment payment = new Payment();
        payment.setTransactionId(txnId);
        payment.setRideId(requestDTO.getRideId());
        payment.setUserId(requestDTO.getUserId());
        payment.setTotalFare(requestDTO.getTotalFare());
        payment.setDiscountAmount(discount);
        payment.setNetAmount(netAmount);
        payment.setPaymentMode(requestDTO.getPaymentMode());
        payment.setPaymentStatus(PaymentStatus.PROCESSING);

        payment = paymentRepository.save(payment);

        // 4. Save initial Audit Log
        auditLogRepository.save(new PaymentAuditLog(
                payment.getPaymentId(),
                PaymentStatus.PENDING,
                PaymentStatus.PROCESSING,
                "Initiating " + requestDTO.getPaymentMode() + " payment"
        ));

        // 5. Select strategy & process payment
        PaymentProcessorStrategy processor = processorFactory.getProcessor(requestDTO.getPaymentMode());
        try {
            payment = processor.process(payment, requestDTO);
        } catch (Exception ex) {
            payment.setPaymentStatus(PaymentStatus.FAILED);
            payment.setFailureReason(ex.getMessage());
            paymentRepository.save(payment);

            auditLogRepository.save(new PaymentAuditLog(
                    payment.getPaymentId(),
                    PaymentStatus.PROCESSING,
                    PaymentStatus.FAILED,
                    "Payment failed: " + ex.getMessage()
            ));
            throw ex;
        }

        payment = paymentRepository.save(payment);

        // 6. Save final Audit Log
        auditLogRepository.save(new PaymentAuditLog(
                payment.getPaymentId(),
                PaymentStatus.PROCESSING,
                PaymentStatus.SUCCESS,
                "Payment successfully processed via " + requestDTO.getPaymentMode()
        ));

        return mapToDTO(payment);
    }

    public PaymentResponseDTO getPaymentById(Integer paymentId) {
        Payment payment = paymentRepository.findById(paymentId)
                .orElseThrow(() -> new PaymentNotFoundException("Payment not found with ID: " + paymentId));
        return mapToDTO(payment);
    }

    public PaymentResponseDTO getPaymentByTransactionId(String transactionId) {
        Payment payment = paymentRepository.findByTransactionId(transactionId)
                .orElseThrow(() -> new PaymentNotFoundException("Payment not found with Transaction ID: " + transactionId));
        return mapToDTO(payment);
    }

    public PaymentResponseDTO getPaymentByRideId(Integer rideId) {
        Payment payment = paymentRepository.findByRideId(rideId)
                .orElseThrow(() -> new PaymentNotFoundException("Payment not found for Ride ID: " + rideId));
        return mapToDTO(payment);
    }

    public List<PaymentResponseDTO> getPaymentsByUserId(Integer userId) {
        return paymentRepository.findByUserIdOrderByCreatedAtDesc(userId)
                .stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    public List<PaymentResponseDTO> getAllPayments() {
        return paymentRepository.findAll()
                .stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    @Transactional
    public PaymentResponseDTO refundPayment(RefundRequestDTO requestDTO) {
        Payment payment = paymentRepository.findById(requestDTO.getPaymentId())
                .orElseThrow(() -> new PaymentNotFoundException("Payment not found with ID: " + requestDTO.getPaymentId()));

        if (payment.getPaymentStatus() != PaymentStatus.SUCCESS) {
            throw new PaymentException("Only successful payments can be refunded. Current status: " + payment.getPaymentStatus());
        }

        PaymentStatus previousStatus = payment.getPaymentStatus();
        PaymentProcessorStrategy processor = processorFactory.getProcessor(payment.getPaymentMode());

        payment = processor.refund(payment, requestDTO.getRefundAmount(), requestDTO.getReason());
        payment = paymentRepository.save(payment);

        auditLogRepository.save(new PaymentAuditLog(
                payment.getPaymentId(),
                previousStatus,
                PaymentStatus.REFUNDED,
                "Refund issued: " + requestDTO.getReason()
        ));

        return mapToDTO(payment);
    }

    public FareCalculationResponseDTO calculateFare(FareCalculationRequestDTO requestDTO) {
        BigDecimal distance = requestDTO.getDistanceKm();
        BigDecimal base = defaultBaseFare != null ? defaultBaseFare : new BigDecimal("50.00");
        BigDecimal perKm = defaultPerKmRate != null ? defaultPerKmRate : new BigDecimal("15.00");

        if (requestDTO.getVehicleTypeName() != null) {
            String type = requestDTO.getVehicleTypeName().toUpperCase();
            if (type.contains("SUV") || type.contains("PREMIUM")) {
                base = new BigDecimal("80.00");
                perKm = new BigDecimal("22.00");
            } else if (type.contains("BIKE") || type.contains("TWO_WHEELER")) {
                base = new BigDecimal("25.00");
                perKm = new BigDecimal("8.00");
            } else if (type.contains("AUTO")) {
                base = new BigDecimal("35.00");
                perKm = new BigDecimal("12.00");
            }
        }

        BigDecimal surge = requestDTO.getSurgeMultiplier() != null && requestDTO.getSurgeMultiplier().compareTo(BigDecimal.ZERO) > 0
                ? requestDTO.getSurgeMultiplier()
                : BigDecimal.ONE;

        BigDecimal distanceFare = distance.multiply(perKm);
        BigDecimal subtotal = base.add(distanceFare);
        BigDecimal totalFare = subtotal.multiply(surge).setScale(2, RoundingMode.HALF_UP);

        FareCalculationResponseDTO response = new FareCalculationResponseDTO();
        response.setDistanceKm(distance);
        response.setBaseFare(base);
        response.setPerKmRate(perKm);
        response.setDistanceFare(distanceFare.setScale(2, RoundingMode.HALF_UP));
        response.setSurgeMultiplier(surge);
        response.setTotalCalculatedFare(totalFare);
        response.setVehicleTypeName(requestDTO.getVehicleTypeName() != null ? requestDTO.getVehicleTypeName() : "STANDARD");

        return response;
    }

    private PaymentResponseDTO mapToDTO(Payment payment) {
        PaymentResponseDTO dto = new PaymentResponseDTO();
        dto.setPaymentId(payment.getPaymentId());
        dto.setTransactionId(payment.getTransactionId());
        dto.setRideId(payment.getRideId());
        dto.setUserId(payment.getUserId());
        dto.setTotalFare(payment.getTotalFare());
        dto.setDiscountAmount(payment.getDiscountAmount());
        dto.setNetAmount(payment.getNetAmount());
        dto.setPaymentMode(payment.getPaymentMode());
        dto.setPaymentStatus(payment.getPaymentStatus());
        dto.setGatewayRef(payment.getGatewayRef());
        dto.setFailureReason(payment.getFailureReason());
        dto.setCreatedAt(payment.getCreatedAt());
        dto.setUpdatedAt(payment.getUpdatedAt());
        return dto;
    }
}

File: app/backend/.net/SmartRide.PaymentService/Services/PaymentService.cs
using Microsoft.EntityFrameworkCore;
using SmartRide.PaymentService.Data;
using SmartRide.PaymentService.Exceptions;
using SmartRide.PaymentService.Models.DTOs;
using SmartRide.PaymentService.Models.Entities;
using SmartRide.PaymentService.Models.Enums;
using SmartRide.PaymentService.Strategies;

namespace SmartRide.PaymentService.Services;

public class PaymentService : IPaymentService
{
    private readonly PaymentDbContext _dbContext;
    private readonly PaymentProcessorFactory _processorFactory;
    private readonly decimal _defaultBaseFare;
    private readonly decimal _defaultPerKmRate;

    public PaymentService(
        PaymentDbContext dbContext,
        PaymentProcessorFactory processorFactory,
        IConfiguration configuration)
    {
        _dbContext = dbContext;
        _processorFactory = processorFactory;
        _defaultBaseFare = configuration.GetValue<decimal>("Fare:DefaultBase", 50.00m);
        _defaultPerKmRate = configuration.GetValue<decimal>("Fare:DefaultPerKm", 15.00m);
    }

    public async Task<PaymentResponseDTO> ProcessPaymentAsync(PaymentRequestDTO requestDTO)
    {
        // 1. Idempotency Check for Ride
        bool hasSuccessfulPayment = await _dbContext.Payments
            .AnyAsync(p => p.RideId == requestDTO.RideId && p.PaymentStatus == PaymentStatus.SUCCESS);
        if (hasSuccessfulPayment)
        {
            throw new DuplicateTransactionException($"Payment has already been completed for Ride #{requestDTO.RideId}");
        }

        // 2. Prepare Transaction ID
        string? txnId = requestDTO.TransactionId;
        if (string.IsNullOrWhiteSpace(txnId))
        {
            txnId = "TXN_" + DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + "_" + Guid.NewGuid().ToString("N")[..6].ToUpper();
        }
        else
        {
            bool txnExists = await _dbContext.Payments.AnyAsync(p => p.TransactionId == txnId);
            if (txnExists)
            {
                throw new DuplicateTransactionException($"Transaction ID already exists: {txnId}");
            }
        }

        // 3. Construct Payment entity
        decimal discount = requestDTO.DiscountAmount ?? 0m;
        decimal netAmount = requestDTO.TotalFare - discount;
        if (netAmount < 0m)
        {
            netAmount = 0m;
        }

        var payment = new Payment
        {
            TransactionId = txnId,
            RideId = requestDTO.RideId,
            UserId = requestDTO.UserId,
            TotalFare = requestDTO.TotalFare,
            DiscountAmount = discount,
            NetAmount = netAmount,
            PaymentMode = requestDTO.PaymentMode,
            PaymentStatus = PaymentStatus.PROCESSING,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        _dbContext.Payments.Add(payment);
        await _dbContext.SaveChangesAsync();

        // 4. Save initial Audit Log
        _dbContext.PaymentAuditLogs.Add(new PaymentAuditLog(
            payment.PaymentId,
            PaymentStatus.PENDING,
            PaymentStatus.PROCESSING,
            $"Initiating {requestDTO.PaymentMode} payment"
        ));
        await _dbContext.SaveChangesAsync();

        // 5. Select strategy & process payment
        var processor = _processorFactory.GetProcessor(requestDTO.PaymentMode);
        try
        {
            payment = await processor.ProcessAsync(payment, requestDTO);
        }
        catch (Exception ex)
        {
            payment.PaymentStatus = PaymentStatus.FAILED;
            payment.FailureReason = ex.Message;
            payment.UpdatedAt = DateTime.UtcNow;
            _dbContext.Payments.Update(payment);

            _dbContext.PaymentAuditLogs.Add(new PaymentAuditLog(
                payment.PaymentId,
                PaymentStatus.PROCESSING,
                PaymentStatus.FAILED,
                $"Payment failed: {ex.Message}"
            ));
            await _dbContext.SaveChangesAsync();
            throw;
        }

        payment.UpdatedAt = DateTime.UtcNow;
        _dbContext.Payments.Update(payment);

        // 6. Save final Audit Log
        _dbContext.PaymentAuditLogs.Add(new PaymentAuditLog(
            payment.PaymentId,
            PaymentStatus.PROCESSING,
            PaymentStatus.SUCCESS,
            $"Payment successfully processed via {requestDTO.PaymentMode}"
        ));
        await _dbContext.SaveChangesAsync();

        return MapToDTO(payment);
    }

    public async Task<PaymentResponseDTO> GetPaymentByIdAsync(int paymentId)
    {
        var payment = await _dbContext.Payments.FindAsync(paymentId)
            ?? throw new PaymentNotFoundException($"Payment not found with ID: {paymentId}");
        return MapToDTO(payment);
    }

    public async Task<PaymentResponseDTO> GetPaymentByTransactionIdAsync(string transactionId)
    {
        var payment = await _dbContext.Payments.FirstOrDefaultAsync(p => p.TransactionId == transactionId)
            ?? throw new PaymentNotFoundException($"Payment not found with Transaction ID: {transactionId}");
        return MapToDTO(payment);
    }

    public async Task<PaymentResponseDTO> GetPaymentByRideIdAsync(int rideId)
    {
        var payment = await _dbContext.Payments.FirstOrDefaultAsync(p => p.RideId == rideId)
            ?? throw new PaymentNotFoundException($"Payment not found for Ride ID: {rideId}");
        return MapToDTO(payment);
    }

    public async Task<List<PaymentResponseDTO>> GetPaymentsByUserIdAsync(int userId)
    {
        var payments = await _dbContext.Payments
            .Where(p => p.UserId == userId)
            .OrderByDescending(p => p.CreatedAt)
            .ToListAsync();
        return payments.Select(MapToDTO).ToList();
    }

    public async Task<List<PaymentResponseDTO>> GetAllPaymentsAsync()
    {
        var payments = await _dbContext.Payments.ToListAsync();
        return payments.Select(MapToDTO).ToList();
    }

    public async Task<PaymentResponseDTO> RefundPaymentAsync(RefundRequestDTO requestDTO)
    {
        var payment = await _dbContext.Payments.FindAsync(requestDTO.PaymentId)
            ?? throw new PaymentNotFoundException($"Payment not found with ID: {requestDTO.PaymentId}");

        if (payment.PaymentStatus != PaymentStatus.SUCCESS)
        {
            throw new PaymentException($"Only successful payments can be refunded. Current status: {payment.PaymentStatus}");
        }

        var previousStatus = payment.PaymentStatus;
        var processor = _processorFactory.GetProcessor(payment.PaymentMode);

        payment = await processor.RefundAsync(payment, requestDTO.RefundAmount, requestDTO.Reason);
        payment.UpdatedAt = DateTime.UtcNow;
        _dbContext.Payments.Update(payment);

        _dbContext.PaymentAuditLogs.Add(new PaymentAuditLog(
            payment.PaymentId,
            previousStatus,
            PaymentStatus.REFUNDED,
            $"Refund issued: {requestDTO.Reason}"
        ));
        await _dbContext.SaveChangesAsync();

        return MapToDTO(payment);
    }

    public FareCalculationResponseDTO CalculateFare(FareCalculationRequestDTO requestDTO)
    {
        decimal distance = requestDTO.DistanceKm;
        decimal baseFare = _defaultBaseFare;
        decimal perKm = _defaultPerKmRate;

        if (!string.IsNullOrWhiteSpace(requestDTO.VehicleTypeName))
        {
            string type = requestDTO.VehicleTypeName.ToUpperInvariant();
            if (type.Contains("SUV") || type.Contains("PREMIUM"))
            {
                baseFare = 80.00m;
                perKm = 22.00m;
            }
            else if (type.Contains("BIKE") || type.Contains("TWO_WHEELER"))
            {
                baseFare = 25.00m;
                perKm = 8.00m;
            }
            else if (type.Contains("AUTO"))
            {
                baseFare = 35.00m;
                perKm = 12.00m;
            }
        }

        decimal surge = requestDTO.SurgeMultiplier > 0m ? requestDTO.SurgeMultiplier : 1.0m;
        decimal distanceFare = Math.Round(distance * perKm, 2, MidpointRounding.AwayFromZero);
        decimal subtotal = baseFare + distanceFare;
        decimal totalFare = Math.Round(subtotal * surge, 2, MidpointRounding.AwayFromZero);

        return new FareCalculationResponseDTO
        {
            DistanceKm = distance,
            BaseFare = baseFare,
            PerKmRate = perKm,
            DistanceFare = distanceFare,
            SurgeMultiplier = surge,
            TotalCalculatedFare = totalFare,
            VehicleTypeName = !string.IsNullOrWhiteSpace(requestDTO.VehicleTypeName) ? requestDTO.VehicleTypeName : "STANDARD"
        };
    }

    private static PaymentResponseDTO MapToDTO(Payment payment)
    {
        return new PaymentResponseDTO
        {
            PaymentId = payment.PaymentId,
            TransactionId = payment.TransactionId,
            RideId = payment.RideId,
            UserId = payment.UserId,
            TotalFare = payment.TotalFare,
            DiscountAmount = payment.DiscountAmount,
            NetAmount = payment.NetAmount,
            PaymentMode = payment.PaymentMode,
            PaymentStatus = payment.PaymentStatus,
            GatewayRef = payment.GatewayRef,
            FailureReason = payment.FailureReason,
            CreatedAt = payment.CreatedAt,
            UpdatedAt = payment.UpdatedAt
        };
    }
}
